python/pyspark/sql/pandas/serializers.py

```diff
@@ -183,30 +193,62 @@ def create_array(s, t):
 
-                # Input partition and result pandas.DataFrame empty, make empty Arrays with struct
-                if len(s) == 0 and len(s.columns) == 0:
-                    arrs_names = [(pa.array([], type=field.type), field.name) for field in t]
+            if self._assign_cols_by_name and any(isinstance(name, str) for name in s.columns):
                 # Assign result columns by schema name if user labeled with strings
-                elif self._assign_cols_by_name and any(isinstance(name, str)
-                                                       for name in s.columns):
-                    arrs_names = [(create_array(s[field.name], field.type), field.name)
-                                  for field in t]
+                by_field_name = True
+            else:
                 # Assign result columns by  position
+                by_field_name = False
+
+            if dt is None:
+                if by_field_name:
+                    return [(create_array(s[field.name], field.type), field.name) for field in t]
+                else:
+                    return [
+                        (create_array(s[s.columns[i]], field.type), field.name)
+                        for i, field in enumerate(t)
+                    ]
+            else:
+                if by_field_name:
+                    return [
+                        (create_array(s[field.name], field.type, struct_field.dataType), field.name)
+                        for field, struct_field in zip(t, dt.fields)
+                    ]
                 else:
-                    arrs_names = [(create_array(s[s.columns[i]], field.type), field.name)
-                                  for i, field in enumerate(t)]
+                    return [
+                        (create_array(s[s.columns[i]], field.type, struct_field.dataType),
+                         field.name)
+                        for i, (field, struct_field) in enumerate(zip(t, dt.fields))
+                    ]
 
+        # Make input conform to [(series1, type1), (series2, type2), ...]
+        if not isinstance(series, (list, tuple)) or \
+                (len(series) == 2 and isinstance(series[1], (pa.DataType, DataType))):
+            series = [series]
+        series = ((s, None) if not isinstance(s, (list, tuple)) else s for s in series)
+
+        arrs = []
+        for s, dt in series:
+            t = to_arrow_type(dt) if isinstance(dt, DataType) else dt
+            if t is not None and pa.types.is_struct(t) and not isinstance(dt, UserDefinedType):
+                if not isinstance(s, pd.DataFrame):
+                    raise ValueError("A field of type StructType expects a pandas.DataFrame, "
+                                     "but got: %s" % str(type(s)))
+                if isinstance(dt, DataType):
+                    type_not_match = "dt must be instance of StructType when t is pyarrow struct"
+                    assert isinstance(dt, StructType), type_not_match
```

Comment:
"dt must be StructType as t is pyarrow struct"

Remove temp variable `type_not_match` with the above shortened error msg. (will change it after the next round of code review)

Reply:
done